benchmark/profile_throughput.py

```diff
@@ -0,0 +1,139 @@
+        if prompt_len < 4 or output_len < 4:
+            # Prune too short sequences.
+            continue
+        if prompt_len > 1024 or prompt_len + output_len > 2048:
+            # Prune too long sequences.
+            continue
+        filtered_dataset.append((prompt, prompt_len, output_len))
+
+    # Sample the requests.
+    sampled_requests = random.sample(filtered_dataset, num_requests)
+    return sampled_requests
+
+
+class Engine:
+
+    def __init__(self, model_path: str, model_name: str):
+        tokenizer_model_path = osp.join(model_path, 'triton_models',
+                                        'tokenizer')
+        tokenizer = Tokenizer(tokenizer_model_path)
+        model = MODELS.get(model_name)()
+        stop_words = model.stop_words
+        tm_model = TurboMind(model_path=model_path, stop_words=stop_words)
+        self.tm_model = tm_model
+        self.tokenizer = tokenizer
+
+    def _inference(self, queue, session_id: int):
+
+        model_inst = self.tm_model.create_instance()
+        while True:
+            request = queue.get()
+            if request is None:
+                # stop signal
+                queue.put(None)
+                return
+            else:
+                prompt, _, output_seqlen = request
+                input_ids = self.tokenizer.encode(prompt)
+
+                for outputs in model_inst.stream_infer(
+                        session_id,
+                        input_ids=input_ids,
+                        request_output_len=output_seqlen,
+                        temperature=1.0,
+                        top_p=1.0,
+                        sequence_start=True,
+                        sequence_end=True,
+                        ignore_eos=True):
+                    res, tokens = outputs[0]
+                    self.tokenizer.decode(res)
+
+    def process_request(self, requests, concurrency: int = 1):
+        q = Queue()
+        threads = []
+
+        start = time.time()
+
+        # start threads
+        for i in range(concurrency):
+            t = Thread(target=self._inference, args=(q, i))
+            t.start()
+            threads.append(t)
+
+        # feed request to q
+        for req in requests:
+            q.put(req)
+
+        q.put(None)
+
+        # wait for finish
+        for t in threads:
+            t.join()
+
+        end = time.time()
+
+        return end - start
+
+
+def main(dataset: str,
+         model_path: str,
+         model_name: str,
+         concurrency: int = 1,
+         num_prompts: int = 1000):
+
+    engine = Engine(model_path, model_name)
+    tokenizer = engine.tokenizer
+
+    requests = sample_requests(dataset, num_prompts, tokenizer)
+
+    elapsed_time = engine.process_request(requests, concurrency)
+    total_num_tokens = sum(prompt_len + output_len
+                           for _, prompt_len, output_len in requests)
+    print(f'Throughput: {len(requests) / elapsed_time:.2f} requests/s, '
```

Comment:
Shall we also output throughput including only generated tokens, which is `sum(output_len) / elapsed_time`?

Reply:
May update the benchmark in readme. And provide another benchmark about turbomind vs vllm on req/min